src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+            CDataTreeView *tree = reinterpret_cast<CDataTreeView *>(frame->m_splitter.GetPane(0, 0));
+
+            if (tree != nullptr) {
+                tree->AddRenderObjects();
+            }
+        }
+
+        if (view != nullptr) {
+            view->EnableRendering(true);
+        }
+
+        return TRUE;
+    }
+
+    return result;
+}
+
+void CW3DViewDoc::Free()
+{
+    if (m_textureScene != nullptr) {
+        if (m_backgroundBitmap != nullptr) {
+            m_backgroundBitmap->Remove();
+        }
+
+        Ref_Ptr_Release(m_textureScene);
+    }
+
+    if (m_backgroundScene != nullptr) {
+        if (m_backgroundObject != nullptr) {
+            m_backgroundObject->Remove();
+        }
+
+        Ref_Ptr_Release(m_backgroundScene);
+    }
+
+    if (m_scene != nullptr) {
+        if (m_model != nullptr) {
+            RemoveRenderObject(m_model);
+        }
+
+        if (m_light != nullptr) {
+            RemoveRenderObject(m_light);
+        }
+
+        m_scene->Remove_All_LOD_Objects();
+        Ref_Ptr_Release(m_scene);
+    }
+
+    Ref_Ptr_Release(m_textureCamera);
+    Ref_Ptr_Release(m_docCamera);
+    Ref_Ptr_Release(m_backgroundBitmap);
+    Ref_Ptr_Release(m_light);
+
+    if (m_model != nullptr) {
+        Ref_Ptr_Release(m_animation);
+        Ref_Ptr_Release(m_model);
+    }
+}
+
+void CW3DViewDoc::RemoveRenderObject(RenderObjClass *robj)
+{
+    if (robj == nullptr) {
+        robj = m_model;
+    }
+
+    while (robj->Get_Num_Sub_Objects() > 0) {
+        RenderObjClass *o = robj->Get_Sub_Object(0);
+
+        if (o != nullptr) {
+            RemoveRenderObject(o);
+            Ref_Ptr_Release(o);
+        }
+    }
+
+    if (robj != nullptr && robj->Class_ID() == RenderObjClass::CLASSID_PARTICLEEMITTER) {
+        ParticleEmitterClass *emitter = reinterpret_cast<ParticleEmitterClass *>(robj);
+        emitter->Stop();
+        emitter->Remove_Buffer_From_Scene();
+    }
+
+    if (m_scene != nullptr) {
+        robj->Remove();
+    }
+}
+
+void CW3DViewDoc::LoadFile(CString pathName)
+{
+    if (m_scene == nullptr) {
+        Initialize();
+    }
+
+    const char *filename = pathName;
+
+    m_filePath = GetFilePath(pathName);
+    m_loadedFiles.Add(pathName);
+
+    CGraphicView *view = GetCurrentGraphicView();
+
+    if (view != nullptr) {
+        view->EnableRendering(false);
+    }
+
+    if (strrchr(filename, '\\') != nullptr) {
+        CString str = filename;
+        str = str.Left(strrchr(filename, '\\') - filename);
+        SetCurrentDirectory(str);
+        g_theSimpleFileFactory->Append_Sub_Directory(str);
+    }
+
+    const char *c = strrchr(filename, '.');
+
+    if (lstrcmpi(c, ".tga") != 0 && lstrcmpi(c, ".dds") != 0) {
+        W3DAssetManager::Get_Instance()->Load_3D_Assets(GetFilenameFromPath(filename));
+    } else {
+        TextureClass *texture = W3DAssetManager::Get_Instance()->Get_Texture(GetFilenameFromPath(filename));
+        Ref_Ptr_Release(texture);

```

Comment:
What is this for? Gets texture and does nothing with it?

Reply:
I don't know why ww did that but they did.